splunk python wrapper, unified direct ingest configuration (#12)
signalfx/splunk-otel-lambda: README.md

Currently, following components are supported:
- [Java wrapper](#java-wrapper) 
- [Python wrapper](#python-wrapper)

### Configuration

All wrapper layers are preconfigured to use Splunk as the tracing backend.

1. Context propagation

    By default, W3C Trace Context and Baggage (`tracecontext,baggage`) propagators are used. 
    
    If you want to change this, set the `OTEL_PROPAGATORS` environment variable in your Lambda function code. 
   
2. Traces export

>     By default, the Jaeger/Thrift modified by Splunk (`jaeger-thrift-splunk`) exporter is used. Endpoint (`OTEL_EXPORTER_JAEGER_ENDPOINT`) is set to `https://ingest.signalfx.com/v2/trace` (Splunk direct ingest). The only thing you need to set is Splunk token:

Reviewer: ```suggestion
    By default, all wrappers use the Jaeger/Thrift exporter modified by Splunk (`jaeger-thrift-splunk`). The endpoint (`OTEL_EXPORTER_JAEGER_ENDPOINT`) is set to `https://ingest.<realm>.signalfx.com/v2/trace`, the direct ingest URL for Splunk Observability Cloud. To make it work, you only need to set the Splunk access token:
```

Reviewer: See https://dev.splunk.com/observability/docs/datamodel/ingest/

     ```

>      SPLUNK_ACCESS_TOKEN: <org access token>

Reviewer: ```suggestion
     SPLUNK_ACCESS_TOKEN: <org_access_token>
```

     ``` 

>     If you want to change the endpoint, set this environment variable in your Lambda function code:

Reviewer: Why would someone want to do this? Can we provide an example? Perhaps an OTel gateway in an EC2 box somewhere?

    ```

>     OTEL_EXPORTER_OTLP_ENDPOINT: <endpoint URL>

Reviewer 2: Shouldn't it be `OTEL_EXPORTER_JAEGER_ENDPOINT`? Since we're using jaeger-thrift by default
```suggestion
    OTEL_EXPORTER_JAEGER_ENDPOINT: <endpoint URL>
```

    ```
    
3. OpenTelemetry Metrics export

>     By default, OpenTelemetry metrics are disabled.    

Reviewer: Why? Can they be enabled? If so, how?

2. Span flush

    Spans are sent synchronously, before the lamda function terminates. Default span flush wait timeout is 1 second. It is controlled with a following environment variable (value in milliseconds):
    ```
    OTEL_INSTRUMENTATION_AWS_LAMBDA_FLUSH_TIMEOUT: 30000
    ```
   
3. Context propagation

    For more information about available context propagators, see the [Propagator settings](https://github.com/open-telemetry/opentelemetry-java/tree/v1.1.0/sdk-extensions/autoconfigure#customizing-the-opentelemetry-sdk) for the OpenTelemetry Java.

4. Logging
    
    Logging is controller with `OTEL_LAMBDA_LOG_LEVEL` environment variable. Default set to `WARN`. When set to `DEBUG`, the logging exporter is added to traces exporter in order to help verify exported data.
    
    Please note that enabling `DEBUG` will generate additional logs which may in turn increase AWS CloudWatch costs. 

### Python wrapper

The Python wrapper is based on Splunk Distribution of OpenTelemetry Python version `v1.0.0`. 

The official documentation of the upstream layer can be found [here](https://github.com/open-telemetry/opentelemetry-lambda/blob/main/python/README.md).

1. Installation

>     No code change is needed to instrument a lambda. However, the `AWS_LAMBDA_EXEC_WRAPPER` environment variable must be set to `/opt/otel-instrument`.

Reviewer: ```suggestion
    Set the `AWS_LAMBDA_EXEC_WRAPPER` environment variable to `/opt/otel-instrument`.
```

Following environment variables should be set:
```
AWS_LAMBDA_EXEC_WRAPPER: /opt/otel-handler

> SPLUNK_ACCESS_TOKEN: <org access token>

Reviewer: ```suggestion
SPLUNK_ACCESS_TOKEN: <org_access_token>
```

```

#### Python

The example assumes the following:

- Data is sent to Splunk APM directly via Ingest endpoint.
- Context propagation is default (`tracecontext,baggage`).
- Realm is `us0`.

Following environment variables should be set:
```
AWS_LAMBDA_EXEC_WRAPPER: /opt/otel-instrument

> SPLUNK_ACCESS_TOKEN: <org access token>

Reviewer: ```suggestion
SPLUNK_ACCESS_TOKEN: <org_access_token>
```
